service/history/hsm/tree.go

```diff
@@ -90,17 +96,90 @@ type cachedMachine struct {
 	deleted bool
 }
 
-type OperationLog struct {
-	// All transitions, including regular state changes and deletions
-	TransitionsByPath map[string][]TransitionOutputWithCount
-	// Track topmost deleted paths to filter outputs efficiently
-	DeletedPaths [][]Key
+// String returns a human-readable representation of a path
+func formatPath(path []Key) string {
+	if len(path) == 0 {
+		return "<root>"
+	}
+	var parts []string
+	for _, key := range path {
+		parts = append(parts, key.String())
+	}
+	return strings.Join(parts, "/")
+}
+
+// Operation represents a state change in the hierarchical state machine tree.
+// Each operation is associated with a path in the tree and provides information
+// about what occurred at that location.
+type Operation interface {
+	// Path returns the full path to the node where this operation occurred
+	Path() []Key
+	mustImplementOperation()
 }
 
-func (o *OperationLog) IsDeleted(path []Key) bool {
-	for _, deletedPath := range o.DeletedPaths {
-		if isPathPrefix(deletedPath, path) {
-			return true
+// DeleteOperation represents the deletion of a node in the tree. When a node
+// is deleted, DeleteOperations are created for it and all its descendants.
+type DeleteOperation struct {
+	// NodePath is the full path to the deleted node
```

Comment:
Can you add punctuation to all docstrings please?

Reply:
Done